trie-db/src/triedbmut.rs

```diff
@@ -1444,17 +1578,27 @@ where
 		match self.storage.destroy(handle) {
 			Stored::New(node) => {
 				let mut k = NibbleVec::new();
-				let encoded_root = node.into_encoded::<_, L::Codec, L::Hash>(
-					|child, o_slice, o_index| {
+
+				let encoded_root = node.into_encoded(
+					|value_node, child, o_slice, o_index| {
 						let mov = k.append_optional_slice_and_nibble(o_slice, o_index);
+						if let Some(value) = value_node {
+							let value_hash = self.db.insert(k.as_prefix(), value);
+							k.drop_lasts(mov);
+							return ChildReference::Hash(value_hash);
+						}
 						let cr = self.commit_child(child, &mut k);
 						k.drop_lasts(mov);
 						cr
-					}
+					},
```

Comment:
this callback function looks quite weird, either value_node is defined and it is supposed to create the childreference of the "value_node" or value_node is none and it is supposed to create the childreference for a regular node.

When calling this in the implementation Value::into_encoded we have to give a fake child parameter if I understand correclty.

IMHO we should create a new type for first argument:
```
enum ChildCreation {
ValueNode(value: &[u8]),
RegularNode(TrieHandle),
}
```
This way I think usage would make more sense